rust/src/lockfile.rs

```diff
@@ -123,36 +135,71 @@ enum LockedPackage {
     },
 }
 
+// NOTE: this is not exactly the same as treefile's merge_map_field
+fn merge_map_field<T>(
+    base: &mut Option<BTreeMap<String, T>>,
+    layer: &mut Option<BTreeMap<String, T>>,
+) {
+    if let Some(layerv) = layer.take() {
+        if let Some(ref mut basev) = base {
+            basev.extend(layerv);
+        } else {
+            *base = Some(layerv);
+        }
+    }
+}
+
 impl LockfileConfig {
-    fn merge(&mut self, other: LockfileConfig) {
-        self.packages.extend(other.packages);
+    fn merge(&mut self, mut other: LockfileConfig) {
+        merge_map_field(&mut self.packages, &mut other.packages);
+        merge_map_field(&mut self.source_packages, &mut other.source_packages);
     }
 
     pub(crate) fn get_locked_packages(&self) -> CxxResult<Vec<crate::ffi::LockedPackage>> {
-        self.packages
-            .iter()
```

Comment:
I think we can avoid the re-indentation by just doing:
```
        self.packages
            .iter()
            .flatten()
```

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=f14ca41d3f0a8ad19b4b0df9fe784fd9

Reply:
Ahh very nice, I'll have to remember this one!

Reply:
Yeah, often people say "I know language X" or "I don't know" but there's this massive spectrum in between around knowing all the bits of the standard libraries, patterns, not to mention crates...

I spent a bit of time studying combinators and how people write them instead of loops and came across `.flatten()` a while ago and successfully internalized it.   I think I found it from this picture:

https://www.reddit.com/r/rust/comments/kdfb9k/rusts_option_in_one_figure/